File: codility_training/lessons.lesson01.Iterations.BinaryGap/junilhwang.js
function solution (N) {
  const b = N.toString(2), len = b.length
  let i = 0, max = 0;
  while (i < len) {
    const idx = b.indexOf('1', i+1) // i 이후의 1을 가져옴
    if (idx === -1) break // 없을 경우 종료
    max = Math.max(idx - i - 1, max) // 현재 거리와 최대 거리를 비교
    i = idx // 다음 1의 위치로 이동
  }
  return max
}

function solution2 (N) {
  return [ ...N.toString(2) ].reduce(([ max, count ], v) => {
    if (v === '0') {
      count += 1
    } else {
      max = Math.max(max, count)
      count = 0
    }
    return [ max, count ]
  }, [0, 0])[0]
}

const assert = require('assert').strict
require('./test.json').forEach(({ input, output }) => {
  assert.deepEqual(solution(...input), output);
  assert.deepEqual(solution2(...input), output);
})

File: codility_training/lessons.lesson01.Iterations.BinaryGap/JunilHwang.java
package lessons.lesson01.Iterations.BinaryGap;

public class JunilHwang {
  public static int solution(int n) {
    String b = Integer.toBinaryString(n);
    int i = 0, max = 0, len = b.length(), idx; // idx는 일단 정의만
    while (i < len) { // 반복 횟수를 최대한 줄이기
      idx = b.indexOf("1", i + 1);
      if (idx == -1) break;
      max = Math.max(max, idx - i - 1);
      i = idx;
    }
    return max;
  }

  public static void main(String[] args) {
    System.out.println(solution(32) == 0);
    System.out.println(solution(1041) == 5);
    System.out.println(solution(9) == 2);
    System.out.println(solution(529) == 4);
    System.out.println(solution(20) == 1);
    System.out.println(solution(15) == 0);
  }
}
